src/app/WriteClient.cpp

```diff
@@ -57,30 +57,45 @@ CHIP_ERROR WriteClient::Init(Messaging::ExchangeManager * apExchangeMgr, Callbac
     return CHIP_NO_ERROR;
 }
 
-void WriteClient::Shutdown()
+void WriteClient::Close()
 {
-    VerifyOrReturn(mState != State::Uninitialized);
-    ClearExistingExchangeContext();
-    ShutdownInternal();
-}
+    MoveToState(State::AwaitingDestruction);
 
-void WriteClient::ShutdownInternal()
-{
-    mMessageWriter.Reset();
+    // OnDone below can destroy us before we unwind all the way back into the
+    // exchange code and it tries to close itself.  Make sure that it doesn't
+    // try to notify us that it's closing, since we will be dead.
+    //
+    // For more details, see #10344.
```

Comment:
I painfully discovered through https://github.com/project-chip/connectedhomeip/issues/13146 and https://github.com/project-chip/connectedhomeip/issues/13237 that when the sdk declares exchange closed, it likely is _not_ done with the exchange context delegate and that the delegate cannot be set null and freed until one more iteration of the chip event loop executes.

Will this actually work here?  When this code signals the app that it is done, can the app immediately free all associated resources?  We need for that to be the case.  Forcing the app to defer cleanup until the event loop comes around for another iteration is not reasonable.

Reply:
@msandstedt I think this is no long a problem after #12794